# Data Cleaning & Quality Check
Verify the Superstore dataset is clean and document any anomalies before EDA.

In [21]:
import sys
sys.path.append("..")

import pandas as pd
import plotly.express as px
from src.data_loader import load_data, clean_data

df_raw = load_data("../data/superstore.csv")
df = clean_data(df_raw)

Shape: (9994, 21)

Column dtypes:
row_id                    int64
order_id                 object
order_date       datetime64[ns]
ship_date        datetime64[ns]
ship_mode                object
customer_id              object
customer_name            object
segment                  object
country                  object
city                     object
state                    object
postal_code               int64
region                   object
product_id               object
category                 object
sub_category             object
product_name             object
sales                   float64
quantity                  int64
discount                float64
profit                  float64
dtype: object

Null counts:
row_id           0
order_id         0
order_date       0
ship_date        0
ship_mode        0
customer_id      0
customer_name    0
segment          0
country          0
city             0
state            0
postal_code      0
region           0
product_id       0


In [22]:
print("Shape:", df.shape)
print("\n--- Sales ---")
print(df["sales"].describe())
print("\n--- Profit ---")
print(df["profit"].describe())
print("\n--- Discount ---")
print(df["discount"].describe())

Shape: (9994, 26)

--- Sales ---
count     9994.000000
mean       229.858001
std        623.245101
min          0.444000
25%         17.280000
50%         54.490000
75%        209.940000
max      22638.480000
Name: sales, dtype: float64

--- Profit ---
count    9994.000000
mean       28.656896
std       234.260108
min     -6599.978000
25%         1.728750
50%         8.666500
75%        29.364000
max      8399.976000
Name: profit, dtype: float64

--- Discount ---
count    9994.000000
mean        0.156203
std         0.206452
min         0.000000
25%         0.000000
50%         0.200000
75%         0.200000
max         0.800000
Name: discount, dtype: float64


In [23]:
out_of_range = df[(df["discount"] < 0) | (df["discount"] > 1)]
print(f"Discount out of [0, 1] range: {len(out_of_range)} rows")

Discount out of [0, 1] range: 0 rows


In [24]:
anomalies = df[(df["profit_margin"] < -50) | (df["profit_margin"] > 100)]
print(f"Profit margin anomalies (<-50% or >100%): {len(anomalies)} rows")
print(anomalies[["product_name", "sales", "profit", "profit_margin", "discount"]].head(10))

Profit margin anomalies (<-50% or >100%): 869 rows
                                         product_name     sales     profit  \
14  Holmes Replacement Filter for HEPA Air Cleaner...    68.810  -123.8580   
15   Storex DuraTech Recycled Plastic Frosted Binders     2.544    -3.8160   
27  Riverside Palais Royal Lawyers Bookcase, Royal...  3083.430 -1665.0522   
28  Avery Recycled Flexi-View Covers for Binding S...     9.618    -7.0532   
32  Acco Pressboard Covers with Storage Hooks, 14 ...     6.858    -5.7150   
36  Electrix Architect's Clamp-On Swing Arm Lamp, ...   190.920  -147.9630   
75                                    Economy Binders     1.248    -1.9344   
76                       6" Cubicle Wall Clock, Black     9.708    -5.8248   
78  Eldon Expressions Desk Accessory, Wood Pencil ...    19.300   -14.4750   
95  Flexible Leather- Look Classic Collection Ring...     5.682    -3.7880   

    profit_margin  discount  
14        -180.00       0.8  
15        -150.00       0.8  


## Summary
- Dataset has 9,994 rows and 26 columns after feature engineering
- Zero null values across all columns
- Discount values are all within valid range [0, 1]
- Some profit margin anomalies exist (heavily discounted items) — these are valid business data points, not errors. No rows removed.
- Data is ready for EDA.

Run this notebook top to bottom and confirm it runs clean. Let me know the output of Cell 5 (how many anomalies) — it'll be interesting and we'll reference it in an insight.

---